src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -689,6 +708,163 @@ MarkNodeHasMetadata(char *nodeName, int32 nodePort, bool hasMetadata)
+		{
+			EnsureSupportedSequenceColumnType(sequenceOid);
+
+			/* create schema if needed */
+			schemaId = get_rel_namespace(sequenceOid);
+			createSchemaCommand = CreateSchemaDDLCommand(schemaId);
+			if (createSchemaCommand != NULL)
+			{
+				sequenceDDLList = lappend(sequenceDDLList, createSchemaCommand);
+			}
+
+			appendStringInfo(wrappedSequenceDef,
+							 WORKER_APPLY_SEQUENCE_COMMAND,
+							 escapedSequenceDef);
+
+			appendStringInfo(sequenceGrantStmt,
+							 "GRANT ALL ON SEQUENCE %s TO %s", sequenceName, ownerName);
+
+			sequenceDDLList = lappend(sequenceDDLList, wrappedSequenceDef->data);
+			sequenceDDLList = lappend(sequenceDDLList, sequenceGrantStmt->data);
+		}
+	}
+
+	return sequenceDDLList;
+}
+
+
+/*
+ * CreateSchemaDDLCommand returns a "CREATE SCHEMA..." SQL string for creating the given
+ * schema if not exists and with proper authorization.
+ */
+char *
+CreateSchemaDDLCommand(Oid schemaId)
+{
+	char *schemaName = get_namespace_name(schemaId);
+	StringInfo schemaNameDef = NULL;
+	char *ownerName = NULL;
+
+	if (strncmp(schemaName, "public", NAMEDATALEN) == 0)
+	{
+		return NULL;
+	}
+
+	schemaNameDef = makeStringInfo();
+	ownerName = SchemaOwner(schemaId);
+	appendStringInfo(schemaNameDef, CREATE_SCHEMA_COMMAND, schemaName, ownerName);
+
+	return schemaNameDef->data;
+}
+
+
+/*
+ * EnsureSupportedSequenceColumnType looks at the column which depends on this sequence
+ * (which it Assert's exists) and makes sure its type is suitable for use in a disributed
+ * manner.
+ *
+ * Any column which depends on a sequence (and will therefore be replicated) but which is
+ * not a BIGINT cannot be used for an mx table, because there aren't enough values to
+ * ensure that generated numbers are globally unique.
+ */
+static void
+EnsureSupportedSequenceColumnType(Oid sequenceOid)
+{
+	Oid tableId = InvalidOid;
+	Oid columnType = InvalidOid;
+	int32 columnId = 0;
+	bool shouldSyncMetadata = false;
+	bool sequenceOwned = false;
+
+	sequenceOwned = sequenceIsOwned(sequenceOid, &tableId, &columnId);
+	if (!sequenceOwned)
+	{
+		aclcheck_error(ACLCHECK_NOT_OWNER, ACL_KIND_SEQUENCE,
+					   get_rel_name(sequenceOid));
+	}
+
+	shouldSyncMetadata = ShouldSyncTableMetadata(tableId);
+
+	columnType = TypeOfColumn(tableId, (int16) columnId);
+
+	if (columnType != INT8OID && shouldSyncMetadata)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot create an mx table with columns which use "
+							   "sequences, but are not BIGINT")));
+	}
+}
+
+
+/*
+ * TypeOfColumn returns the Oid of the type of the provided column of the provided table.
+ */
+static Oid
+TypeOfColumn(Oid tableId, int16 columnId)
+{
+	const int scanKeyCount = 2;
+	ScanKeyData scanKey[scanKeyCount];
+	SysScanDesc scanDescriptor = NULL;
+	Oid returnType = InvalidOid;
+	HeapTuple tup = NULL;
+	Form_pg_attribute attrForm = NULL;
+
+	Relation pgAttribute = heap_open(AttributeRelationId, AccessShareLock);
+
+	ScanKeyInit(&scanKey[0],
```

Comment:
This is great. With this, I have also changed the name of the function since it can give the owner of any class.